CedarJavaFFI/src/interface.rs

```diff
@@ -171,6 +171,145 @@ pub fn json_parse_entity_uid(input: &str) -> InterfaceResult {
+            Ok(_) => Ok(JValueGen::Object(env.new_string("Success")?.into())),
+        }
+    }
+}
+
+#[jni_fn("com.cedarpolicy.model.slice.Policy")]
+pub fn parsePolicyJni<'a>(mut env: JNIEnv<'a>, _: JClass, policy_jstr: JString<'a>) -> jvalue {
+    match parse_policy_internal(&mut env, policy_jstr) {
+        Err(e) => jni_failed(&mut env, e.as_ref()),
+        Ok(policy_text) => policy_text.as_jni(),
+    }
+}
+
+fn parse_policy_internal<'a>(
+    env: &mut JNIEnv<'a>,
+    policy_jstr: JString<'a>,
+) -> Result<JValueOwned<'a>> {
+    if policy_jstr.is_null() {
+        raise_npe(env)
+    } else {
+        let policy_jstring = env.get_string(&policy_jstr)?;
+        let policy_string = String::from(policy_jstring);
+        match Policy::from_str(&policy_string) {
+            Err(e) => Err(Box::new(e)),
+            Ok(p) => {
+                let policy_text = format!("{}", p);
+                Ok(JValueGen::Object(env.new_string(&policy_text)?.into()))
+            }
+        }
+    }
+}
+
+#[jni_fn("com.cedarpolicy.model.slice.Policy")]
+pub fn parsePolicyTemplateJni<'a>(
+    mut env: JNIEnv<'a>,
+    _: JClass,
+    template_jstr: JString<'a>,
+) -> jvalue {
+    match parse_policy_template_internal(&mut env, template_jstr) {
+        Err(e) => jni_failed(&mut env, e.as_ref()),
+        Ok(template_text) => template_text.as_jni(),
+    }
+}
+
+fn parse_policy_template_internal<'a>(
+    env: &mut JNIEnv<'a>,
+    template_jstr: JString<'a>,
+) -> Result<JValueOwned<'a>> {
+    if template_jstr.is_null() {
+        raise_npe(env)
+    } else {
+        let template_jstring = env.get_string(&template_jstr)?;
+        let template_string = String::from(template_jstring);
+        match Template::from_str(&template_string) {
+            Err(e) => Err(Box::new(e)),
+            Ok(template) => {
+                let template_text = template.to_string();
+                Ok(JValueGen::Object(env.new_string(&template_text)?.into()))
+            }
+        }
+    }
+}
+
+#[jni_fn("com.cedarpolicy.model.slice.Policy")]
+pub fn validateTemplateLinkedPolicyJni<'a>(
+    mut env: JNIEnv<'a>,
+    _: JClass,
+    template_jstr: JString<'a>,
+    jprincipal_euid: JObject<'a>,
+    jresource_euid: JObject<'a>,
+) -> jvalue {
+    match validate_template_linked_policy_internal(
+        &mut env,
+        template_jstr,
+        jprincipal_euid,
+        jresource_euid,
+    ) {
+        Err(e) => jni_failed(&mut env, e.as_ref()),
+        Ok(v) => v.as_jni(),
+    }
+}
+
+fn validate_template_linked_policy_internal<'a>(
+    env: &mut JNIEnv<'a>,
+    template_jstr: JString<'a>,
+    jprincipal_euid: JObject<'a>,
+    jresource_euid: JObject<'a>,
+) -> Result<JValueOwned<'a>> {
+    if template_jstr.is_null() {
+        raise_npe(env)
+    } else {
+        let template_jstring = env.get_string(&template_jstr)?;
+        let template_string = String::from(template_jstring);
+        let template = Template::from_str(&template_string)?;
```

Comment:
In particular (re. the exception comment above), should we handle these `Err`s as user errors and return false?

Reply:
This is an artifact of the cross-language boundary. In order to avoid this i'd have to write a whole other series of JPolicy objects representing the rust-side representations of java heap objects.